arch/eventloop_posix.c

```diff
@@ -0,0 +1,693 @@
+    UA_TimerEntry *first = ZIP_MIN(UA_TimerZip, &el->timerRoot);
+    while(first && first->nextTime <= nowMonotonic) {
+
+        processTimerEntry(el, nowMonotonic, first);
+        first = ZIP_MIN(UA_TimerZip, &el->timerRoot);
+    }
+    /* Return the timestamp of the earliest next callback */
+    return (first) ? first->nextTime : UA_INT64_MAX;
+}
+
+static void
+freeTimerEntry(UA_TimerEntry *te, void *data) {
+    UA_free(te);
+}
+
+/* Process and then free registered delayed callbacks */
+static void
+processDelayed(UA_EventLoop *el) {
+    UA_LOCK_ASSERT(&el->elMutex, 1);
+    while(el->delayedCallbacks) {
+        UA_DelayedCallback *dc = el->delayedCallbacks;
+        el->delayedCallbacks = dc->next;
+        /* Delayed Callbacks might have no cb pointer if all we want to do is
+         * free the memory */
+        if(dc->callback) {
+            UA_UNLOCK(&el->elMutex);
+            dc->callback(dc->application, dc->data);
+            UA_LOCK(&el->elMutex);
+        }
+        UA_free(dc);
+    }
+}
+
+/***********************/
+/* EventLoop Lifecycle */
+/***********************/
+
+UA_EventLoop *
+UA_EventLoop_new(const UA_Logger *logger) {
+    UA_EventLoop *el = (UA_EventLoop*)UA_malloc(sizeof(UA_EventLoop));
+    if(!el)
+        return NULL;
+    memset(el, 0, sizeof(UA_EventLoop));
+    UA_LOCK_INIT(&el->elMutex);
+    el->logger = logger;
+    return el;
+}
+
+UA_StatusCode
+UA_EventLoop_delete(UA_EventLoop *el) {
+    UA_LOCK(&el->elMutex);
+
+    /* Check if the EventLoop can be deleted */
+    if(el->state != UA_EVENTLOOPSTATE_STOPPED &&
+       el->state != UA_EVENTLOOPSTATE_FRESH) {
+        UA_LOG_WARNING(el->logger, UA_LOGCATEGORY_EVENTLOOP,
+                       "Cannot delete a running EventLoop");
+        UA_UNLOCK(&el->elMutex);
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    /* Deregister and delete all the EventSources */
+    while(el->eventSources) {
+        UA_EventSource *es = el->eventSources;
+        UA_EventLoop_deregisterEventSource(el, es);
+        es->free(es);
+    }
+
+    /* Remove the repeated timed callbacks */
+    ZIP_ITER(UA_TimerZip, &el->timerRoot, freeTimerEntry, NULL);
+
+    /* Process remaining delayed callbacks */
+    processDelayed(el);
+
+    /* Clean up */
+    UA_UNLOCK(&el->elMutex);
+    UA_LOCK_DESTROY(&el->elMutex);
+    UA_free(el);
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_EventLoopState
+UA_EventLoop_getState(UA_EventLoop *el) {
+    return el->state;
+}
+
+UA_StatusCode
+UA_EventLoop_start(UA_EventLoop *el) {
+    UA_LOCK(&el->elMutex);
+    if(el->state != UA_EVENTLOOPSTATE_FRESH &&
+       el->state != UA_EVENTLOOPSTATE_STOPPED) {
+        UA_UNLOCK(&el->elMutex);
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    UA_LOG_INFO(el->logger, UA_LOGCATEGORY_EVENTLOOP, "Starting the EventLoop");
+
+    UA_EventSource *es = el->eventSources;
+    UA_StatusCode res = UA_STATUSCODE_GOOD;
+    while(es) {
+        res |= es->start(es);
```

Comment:
> The CI is failing for the multi-threaded build.
> 
> Otherwise this looks good now.
> 
> I will make a 1.3 branch for the next stable release.
> Then we can apply this to master.

Great, regrading multithreading: 

- One multithreading test could be fixed by just adding missing `UA_UNLOCK` calls.
- Another test's (`check_eventloop_tcp`) issue is about the lock being a non-reentrantlock.
  - In the snipped above we call `es->start(es);` within `UA_EventLoop_start` with the lock `elMutex` being in a locked state.
  - `es->start` in case of tcp calls `TCP_eventSourceStart`
  - thereby registering the listen socket via `TCP_registerListenSocketDomainName` -> `TCP_registerListenSocket` -> `UA_EventLoop_registerFD`
  - `UA_EventLoop_registerFD` in turn does a call to `UA_LOCK` which is illegal because it is non-reentrant and allready locked by `UA_EventLoop_start`.
  
**Possible Solutions**

1. make the lock reentrant.
2. call `UA_UNLOCK` before `es->start` and `UA_LOCK` afterwards 
```c
UA_UNLOCK(&el->elMutex);
res |= es->start(es);
UA_LOCK(&el->elMutex);
```